modules/proxy/proxy_util.c

```diff
@@ -5053,8 +5085,22 @@ static int proxy_tunnel_transfer(proxy_tunnel_rec *tunnel,
             return HTTP_INTERNAL_SERVER_ERROR;
         }
 
-        del_pollset(tunnel->pollset, in->pfd, APR_POLLIN);
-        add_pollset(tunnel->pollset, out->pfd, APR_POLLOUT);
+        if (out->pfd->desc_type == APR_POLL_SOCKET) {
+            /* if the output is a SOCKET, we can stop polling the input
+             * until the output signals POLLOUT again. */
+            del_pollset(tunnel->pollset, in->pfd, APR_POLLIN);
+            add_pollset(tunnel->pollset, out->pfd, APR_POLLOUT);
+        }
+        else {
+            /* Unable to use POLLOUT in this direction, out is not a socket */
+            if (in->down_in) {
+                del_pollset(tunnel->pollset, in->pfd, APR_POLLIN);
+                ap_log_rerror(APLOG_MARK, APLOG_TRACE3, 0, tunnel->r,
+                              "proxy: %s: %s write shutdown",
+                              tunnel->scheme, out->name);
+                out->down_out = 1;
+            }
```

Comment:
Here I think we should remove POLLIN for in->pfd unconditionally and ap_assert(APR_STATUS_IS_EOF(rv)) for the APR_POLL_FILE case.
IIUC, there is no reason that APR_STATUS_IS_INCOMPLETE(rv) is returned above because h2_c2_filter_out() will block to flush the data if necessary thus it will never grow any buffer indefinitely nor set pending data aside to trigger ap_filter_should_yield().
If so we are good here too, in 2.4.x we are not nonblocking anyway for now.

I would just make the assumptions clear, writing this part like:
```
        del_pollset(tunnel->pollset, in->pfd, APR_POLLIN);
        if (out->pfd->desc_type == APR_POLL_SOCKET) {
            /* If the output is a SOCKET, we stop polling the input
             * until the output signals POLLOUT again.
             */
            add_pollset(tunnel->pollset, out->pfd, APR_POLLOUT);
        }
        else {
            /* We can't use POLLOUT in this direction for the only APR_POLL_FILE case
             * we have so far (mod_h2's "signal" pipe), we assume that the client's
             * ouput filters chain will block/flush if necessary (i.e. no pending data),
             * hence that the origin is EOF when reaching here. This direction is over.
             */
            ap_assert(in->down_in && APR_STATUS_IS_EOF(rv));
            ap_log_rerror(APLOG_MARK, APLOG_TRACE3, 0, tunnel->r,
                          "proxy: %s: %s write shutdown (pipe)",
                          tunnel->scheme, out->name);
            out->down_out = 1;
        }
```

Reply:
Changed to that suggestion. Thanks.